core/src/HIP/Kokkos_HIP_Instance.cpp

```diff
@@ -296,15 +318,18 @@ void HIPInternal::finalize() {
     RecordHIP::decrement(RecordHIP::get_record(m_scratchFlags));
     RecordHIP::decrement(RecordHIP::get_record(m_scratchSpace));
 
-    for (int i = 0; i < m_n_team_scratch; ++i) {
-      if (m_team_scratch_current_size[i] > 0)
-        Kokkos::kokkos_free<Kokkos::HIPSpace>(m_team_scratch_ptr[i]);
-    }
+    if (m_scratchFunctorSize > 0)
+      RecordHIP::decrement(RecordHIP::get_record(m_scratchFunctor));
+  }
```

Comment:
Did you mean to nest in the `if (nullptr != m_scratchSpace || nullptr != m_scratchFlags)` block?